inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/non_max_suppression_gpu_ref.cl

```diff
@@ -0,0 +1,571 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "include/data_types.cl"
+#include "include/fetch.cl"
+
+/// optional input variables
+/// NUM_SELECT_PER_CLASS_VAL TO_UNIT_TYPE(num_select_per_class[0]),   default is 0 
+/// IOU_THRESHOLD_VAL        TO_ACCUMULATOR_TYPE(iou_threshold[0]),   default is ACCUMULATOR_VAL_ZERO
+/// SCORE_THRESHOLD_VAL      TO_ACCUMULATOR_TYPE(score_threshold[0]), default is ACCUMULATOR_VAL_ZERO
+/// SOFT_NMS_SIGMA_VAL       TO_ACCUMULATOR_TYPE(soft_nms_sigma[0]),  default is ACCUMULATOR_VAL_ZERO
+/// OUTPUT_NUM               Number of outputs. [OUTPUT_NUM, 3, 1, 1]
+/// BUFFER_STRIDE            20 bytes * NUM_BOXES
+
+#define unroll_for __attribute__((opencl_unroll_hint)) for
+
+#define NUM_BATCHES     INPUT0_BATCH_NUM
+#define NUM_BOXES       INPUT0_FEATURE_NUM
+#define NUM_CLASSES     INPUT1_FEATURE_NUM
+
+typedef struct {
+    ushort boxId;
+    int suppress_begin_index;
+    INPUT1_TYPE score;
+} FUNC(SortedBoxInfo);
+
+typedef struct {
+    short batchId;
+    ushort classId;
+    ushort boxId;
+    INPUT1_TYPE score;
+} FUNC(BoxInfo);
+
+#define SBOX_INFO FUNC(SortedBoxInfo)
+#define BOX_INFO FUNC(BoxInfo)
+
+inline float FUNC(intersectionOverUnion)(const __global INPUT0_TYPE *boxes,
+    const short batchA, const ushort boxIdA,
+    const short batchB, const ushort boxIdB)
+{
+    const float4 pA = convert_float4(vload4(0, &boxes[INPUT0_GET_INDEX(batchA, boxIdA, 0, 0)]));
+    const float4 pB = convert_float4(vload4(0, &boxes[INPUT0_GET_INDEX(batchB, boxIdB, 0, 0)]));
+
+#if BOX_ENCODING == 0
+    const float ax1 = min(pA[1], pA[3]);
```

Comment:
It looks that this sorting in frame of one box not required for each IoU. It is proposed to process all the boxes ones to be sure that x[1]<=x[2] and y[1]<=y[2]

Reply:
x1/x2 and y1/y2 should be aligned because the inputs are not aligned in test cases as gpuFuncTests and clDNN_unit_tests.

Reply:
The idea is align them ones for all the boxes (preprocess) in one pass but not in each IoU computation.

Reply:
Sorry, but cannot understand what you are suggesting. Are you saying we need to have a separate pass to make entire boxes to have x1 < x2 condition? If so, it would require additional intermediate buffer..

Reply:
Yes, I propose to measure if preprocess with intermediate buffer will be better than current way

Reply:
Understood your point. It may have some advantage as `min()` and `max()` are much more parallelized, but it will require additional access to global memory. So the overall performance gain is blurry to us at this moment. Let us follow-up when further NMS optimization target becomes clear.